### ISSUE TITLE
[VL] Track Jvm based libhdfs performance gap issue

### ISSUE THREAD

**Issue by JkSelf** [apache/incubator-gluten#7746]
### Description

Currently, we supported JVM-based libhdfs in both the [Velox ](https://github.com/facebookincubator/velox/pull/9835)and [Gluten ](https://github.com/apache/incubator-gluten/pull/6172)communities. However, the performance in Jenkins tests has not met our expectations.

- Static linking with libhdfs3.so: 340%

- Dynamic linking with JVM: 214.75% (requires further investigation)

- Dynamic linking with libhdfs3: 323.85% (only replacing libhdfs.so with libhdfs3.so)

- Expected performance for static linking in [PR#7697](https://github.com/apache/incubator-gluten/pull/7697): 340%

**Comment by FelixYBW** [comment]
FYI, static means link at compile time. Dynamic here means loading so library at runtime.

Ratio above means the performance boost compared to Vanilla Spark.

**Comment by FelixYBW** [comment]
Looks the indirect call to jvm is not efficient.

**Comment by zhanglistar** [comment]
I thought IO is the big head, indirect call is roughly the virtual table cost.

**Comment by FelixYBW** [comment]
All above perf test is on single node with enough SSDs. Not sure if the short circuit is enabled or not in JVM test, but even not the perf shouldn't drop so much.

**Comment by FelixYBW** [comment]
More clear. @JkSelf Can you also test the missing config?

 perf gain over vanilla | link at compile time | load at runtime
-- | -- | --
libhdfs3 | 3.4 | 3.23
jvm libhdfs | | 2.14

**Comment by zhanglistar** [comment]
So you guys turn to static linking?

**Comment by FelixYBW** [comment]
> So you guys turn to static linking?

No, latest perf test shows libhdfs3 has the same perf. If libhdfs jvm also has the same perf, we can just use the runtime loading way.

 perf gain over vanilla | link at compile time | load at runtime
-- | -- | --
libhdfs3 | 3.23 | 3.23
jvm libhdfs | | 2.14

**Comment by FelixYBW** [comment]
Thank you Ke. Can you add a readme and describe how to config libhdfs3 and hadoop libhdfs? Is changing executorEnv.LD_LIBRARY_PATH enough?

Going fwd, we will only support run time load of libhdfs library.

**Comment by Yohahaha** [comment]
I found this issue more severe in our environment.

TPCDS sf1000 + Spark 3.4 + 192core

date 1024: 494s
date 1025: 632s (1.27x)
date 1119: 580s (1.17x)

@FelixYBW

**Comment by JkSelf** [comment]
@Yohahaha `date 1024`, data was read using the native libhdfs3 library, which does not take advantage of short-circuit reads. However, `date 1119`, data was accessed using the JVM-based libhdfs, which benefits from short-circuit reads. I enabled short-circuit reads in my local test, resulting in a 1.09x performance improvement as noted above. And you can re-test with enabling short circuit read.

**Comment by Yohahaha** [comment]
> [@Yohahaha](https://github.com/Yohahaha) `date 1024`, data was read using the native libhdfs3 library, which does not take advantage of short-circuit reads. However, `date 1119`, data was accessed using the JVM-based libhdfs, which benefits from short-circuit reads. I enabled short-circuit reads in my local test, resulting in a 1.09x performance improvement as noted above. And you can re-test with enabling short circuit read.

I have set `dfs.client.read.shortcircuit=true`, are there any configurations I may have missed?

**Comment by JkSelf** [comment]
@Yohahaha You also need to set the `dfs.domain.socket.path` as below.

```
<property>
      <name>dfs.client.read.shortcircuit</name>
      <value>true</value>
    </property>
    <property>
      <name>dfs.domain.socket.path</name>
      <value>/var/lib/hadoop-hdfs/dn_socket</value>
    </property> 
```

**Comment by Yohahaha** [comment]
> [@Yohahaha](https://github.com/Yohahaha) You also need to set the `dfs.domain.socket.path` as below.
> 
> ```
> <property>
>       <name>dfs.client.read.shortcircuit</name>
>       <value>true</value>
>     </property>
>     <property>
>       <name>dfs.domain.socket.path</name>
>       <value>/var/lib/hadoop-hdfs/dn_socket</value>
>     </property> 
> ```

`dfs.domain.socket.path` was set too.

**Comment by JkSelf** [comment]
@Yohahaha No other configurations need to set.
